src/GitHub/Json.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeOperators     #-}
+
+{- |
+Copyright: (c) 2021 Kowainik
+SPDX-License-Identifier: MPL-2.0
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+
+Utilities to parse JSON returned by the API easier.
+-}
+
+module GitHub.Json
+    ( Nested (..)
+    ) where
+
+import Data.Aeson (FromJSON (..), withObject, (.:))
+import Data.Proxy (Proxy (..))
+import Data.Text (Text)
+import GHC.TypeLits (KnownSymbol, Symbol, symbolVal)
+
+import qualified Data.Text as Text
+
+
+{- | GitHub GraphQL API usually returns very nested JSON objects:
+
+@
+{
+  "data": {
+    "repository": {
+      "milestone": {
+        "id": "MDk6TWlsZXN0b25lNDY3NjQwNQ=="
+      }
+    }
+  }
+}
+@
+
+Since @aeson@ doesn't provide easy ways to parse nested objects and we
+don't want to depend on the @lens-aeson@ package, we provide this
+newtype wrapper.
+
+To parse the above JSON into a single value of type 'MilestoneId', use one of
+the values of one the following types:
+
+@
+Nested '[ "data", "repository", "milestone", "id" ] Text
+Nested '[ "data", "repository", "milestone" ] MilestoneId
+@
+
+However, in real-life when using queries of this library, you don't
+need to specify @"data"@ as it handled internally.
+
+The code looks like this:
+
+@
+unNested @'[ "repository", "milestone" ] @MilestoneId <$> callGitHubRaw
+@
+-}
+newtype Nested (keys :: [Symbol]) a = Nested
```

Comment:
This module contains a little bit of type-level magic, but I don't know a better way at the moment 🤔 

This library is becoming more and more complex, I'm sorry for that... It's so awkward to work with GraphQL in Haskell type-safely